### ISSUE TITLE
Create some automation to validate C# compiler with .NET Native

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#18754]
How do we catch integration problems between C# compiler and .NET Native toolchain before they reach users?

Relates to https://github.com/dotnet/roslyn/issues/18737
FYI @jaredpar @agocke for brainstorming

**Comment by jcouv** [comment]
Some brainstorming options:
1. compiling Roslyn to a native exe, and then using that output to compile Roslyn again (there are some known p/invoke issues)
2. introduce a new runner for `CompileAndVerify` in our test suite which would not just (1) emit the compilation to native then (2) run it. Rather the run step would become "compile to native then run". We'd have to consume the .NET Native toolset as a package (which it seems corefx already does).

**Comment by jcouv** [comment]
From discussion with Andy, the second option will ultimately run into the problem that running the native toolset (to compile from IL to native) takes long even for small programs. 
There are intermediate problems that we might be able to work around (introducing a persistent process for loading native bits and using some caching strategy to only send it bits that it doesn't know about), but the engineering is probably not worth it if the run time ends up being prohibitive anyways.
Option (1) still seems more straightforward.

**Comment by jkotas** [comment]
Would option (1) catch the problem from #18737 soon enough?

**Comment by jaredpar** [comment]
@jkotas yes for both this case specifically and other language features in general.  The Roslyn team is quite aggressive about dogfooding our features early.  It's advantageous to both the compiler team and the IDE team hence we actively encourage it.  

The self hosted compiler builds already catch a number of our own mistakes in CI. Using Net Native here to do the same would likely produce similar wins.

**Comment by jcouv** [comment]
@MeiChin-Tsai We'll wait for your decision.

From my understanding, we'd need the following work items:
1. package of .NET Native tools (there is one, used by corefx, but we need to verify it can be used by our automation, in terms of policy and permissions),
2. the .NET Native package should offer an API for IL-to-native conversion without spawning processes (this is not strictly required, but without it, the overhead will make the performance really slow for the volume of tests. Maybe we stage this in phases?),
3. creating a test runner for .NET Native in compiler test framework (in addition to the existing ones for desktop and Core CLR) and hook up to automation,
4. investigate and categorize failures, skip tests, file issues as appropriate. (I'm assuming .NET Native team would own this effort, since the tests already pass on desktop and Core)

**Comment by jkotas** [comment]
> the .NET Native package should offer an API for IL-to-native conversion without spawning processes 

This is not realistic. Spawning multiple processes is pretty fundamental to the design of .NET Native toolchain.

**Comment by MeiChin-Tsai** [comment]
@jcouv - Is not having item 2 a deal breaker?

**Comment by jaredpar** [comment]
It's not a deal breaker.  It's an idea we had for optimizing the testing a bit.

**Comment by jcouv** [comment]
I'm ok to start without it. We'll see how it works in practice. 
Our back-of-the-envelope estimate is 10 days to run all the tests (using Matt Whilden's 17 seconds per test, based on simple hello world).

**Comment by MeiChin-Tsai** [comment]
@yizhang82 - could you work out the logistic of .Net Native toolchain to Roslyn? thx.

**Comment by yizhang82** [comment]
Chatted with @jcouv offline. One thing we had enabled for our N tests in the past is to "merge" multiple tests into one at compile time by building them as one EXE, and send it to ILC. This is very effective for small tests, which is the case for Roslyn tests. In particular, Roslyn has concept of test runners that runs the emitted assembly in the target environment, and in theory we could have N test runner to defer compilation until it has reached a threshold (say 10), and it compile the tests together as one 'app', run the app 10 times with each test scenario, and then get back to the waiting tests to fetch the results. It can also be smart enough to group compatible tests if there are tests are fundamentally incompatible. But with enough tweaks to Roslyn test infra/tests (such as giving assemblies unique name, etc) hopefully we can minimize those incompatible test cases that can't be merged. 
@jcouv is going to look deeper into the test runners and see if the above is feasible.

**Comment by jcouv** [comment]
Another issue that came up recently: https://github.com/dotnet/corefx/issues/22903
